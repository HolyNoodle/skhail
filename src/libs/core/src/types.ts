import { SkhailService } from "./classes/Service";

export type ErrorCode = "denied" | "unexpected" | "not_found";

export interface ISkhailError<ErrorType extends string = ErrorCode> {
  name: ErrorType;
  message: string;
  details?: any;
  stack?: string;
}

export enum LogLevel {
  DEBUG,
  TRACE,
  INFO,
  WARNING,
  ERROR,
}

export type EmitterListener = (...args: any[]) => Promise<void>;

export interface IEmitter extends ILifeCycle {
  setLogger(logger: ILogger): void;
  on(group: string, event: string, listener: EmitterListener): Promise<void>;
  off(group: string, event: string, listener: EmitterListener): Promise<void>;
  emit(event: string, args: any[]): Promise<void>;
}

export interface LogMessage {
  level: LogLevel;
  tid?: string;
  message: string;
  instance: string;
  scope: string;
  details?: Object & any;
  date: number;
}

export interface ILogger extends ILifeCycle {
  getInstance(): string;
  clone(): ILogger;
  setInstance(instance: string): void;
  setScope(scope: string): void;
  setTransactionId(tid: string): void;
  log(message: LogMessage): void;
  trace(message: string, details?: {}): void;
  debug(message: string, details?: {}): void;
  info(message: string, details?: {}): void;
  warning(message: string, details?: {}): void;
  error(message: string, details?: {}): void;
}

export interface IEnveloppe<Context extends ContextOptions> {
  service: string;
  method: string;
  args: any[];
  context: IRequestContext<Context>;
}

export type IEnveloppeResponse =
  | { tid: string } & (
      | {
          success: true;
          response?: any;
        }
      | {
          success: false;
          error: ISkhailError;
        }
    );

export type EnveloppeHandler<Context extends ContextOptions> = (
  enveloppe: IEnveloppe<Context>
) => Promise<IEnveloppeResponse>;

export type EventOptions = {
  [key: string]: any[];
};

export type ContextOptions = {
  [key: string]: any;
};

export interface IQueue<Context extends ContextOptions> extends ILifeCycle {
  setLogger(logger: ILogger): void;
  setHandler(
    service: string,
    handler: EnveloppeHandler<Context>
  ): Promise<void>;
  enqueue(enveloppe: IEnveloppe<Context>): Promise<IEnveloppeResponse>;
}
export interface QueueWrapper<ContextType extends ContextOptions> {
  get<Service extends SkhailService<ContextType>>(
    type: Constructor<Service>,
    context?: ContextType,
    forwardedContext?: IRequestContext<ContextType>
  ): ServiceFunctions<Service, ContextType>;
}
export interface ISkhailClient<ContextType extends ContextOptions>
  extends QueueWrapper<ContextType> {
  start(): Promise<void>;
  stop(): Promise<void>;
}
export interface ISkhailServer<ContextType extends ContextOptions> {
  start(): Promise<void>;
  stop(): Promise<void>;

  get<
    Service extends SkhailService<ContextType, EventType>,
    EventType extends EventOptions
  >(
    type: Constructor<Service>,
    context?: ContextType,
    forwardedContext?: IRequestContext<ContextType>
  ): ServiceFunctions<Service, ContextType>;

  getQueue(): IQueue<ContextType>;
  getEmitter(): IEmitter | undefined;
  getClient(): ISkhailClient<ContextType>;
}

export interface Middleware<
  ContextIn extends ContextOptions,
  ContextOut extends ContextOptions = ContextIn
> {
  id: string;
  prepare?: (service: SkhailService<ContextIn>) => Promise<void>;
  cleanup?: (service: SkhailService<ContextIn>) => Promise<void>;
  process: (
    enveloppe: IEnveloppe<ContextIn>,
    service: SkhailService<ContextOut>
  ) => Promise<ContextOut | undefined>;
}

export type ServiceClassDescriptor = {
  identifier: string;
  getMiddlewares: () => Middleware<any, any>[];
  isManaged: () => boolean;
};
export type AbstractConstructor<
  Service = any,
  Args extends any[] = any[]
> = (abstract new (...args: Args) => Service) & ServiceClassDescriptor;
export type DefinedConstructor<
  Service = any,
  Args extends any[] = any[]
> = (new (...args: Args) => Service) & ServiceClassDescriptor;
export type Constructor<Service = any, Args extends any[] = any[]> =
  | AbstractConstructor<Service, Args>
  | DefinedConstructor<Service, Args>;

export interface IRequestContext<Context extends ContextOptions> {
  tid: string;
  origin?: string;
  data?: Context;
  parent?: IRequestContext<Context>;
}

export type EventFunction<EventDataType extends Array<any>> = (
  ...args: EventDataType
) => Promise<void> | void;

export type ServiceFunction = (...args: any[]) => Promise<any>;
export type ServiceFunctions<
  Service extends SkhailService<ContextType>,
  ContextType extends ContextOptions
> = {
  [Key in keyof Service as Service[Key] extends ServiceFunction
    ? Key
    : never]: Service[Key] extends ServiceFunction ? Service[Key] : never;
};

export abstract class ILifeCycle {
  abstract prepare?(): Promise<void>;
  abstract ready?(): Promise<void>;
  abstract cleanup?(): Promise<void>;
}

export abstract class IEventEmitter<EventType extends EventOptions> {
  abstract emit<Event extends keyof EventType>(
    event: Event,
    args: EventType[Event]
  ): void;
}
export abstract class IEventListener {
  abstract on<
    Service extends SkhailService<any, EventType>,
    EventType extends EventOptions,
    Event extends keyof EventType
  >(
    service: Constructor<Service>,
    event: Event,
    listener: EventFunction<EventType[Event]>
  ): void;
  abstract off<
    Service extends SkhailService<any, EventType>,
    EventType extends EventOptions,
    Event extends keyof EventType
  >(
    service: Constructor<Service>,
    event: Event,
    listener: EventFunction<EventType[Event]>
  ): void;
}
