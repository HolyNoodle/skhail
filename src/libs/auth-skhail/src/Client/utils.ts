import { KeyUtils } from "@skhail/auth";

export const getPrivateKey = async ({
  privateKey,
  passphrase,
}: {
  privateKey?: KeyUtils.Key;
  passphrase?: string;
}) => {
  if (passphrase && !privateKey) {
    const store = KeyUtils.createNewKeyStore();
    return await KeyUtils.createNewPrivateKey(store, passphrase);
  }

  return privateKey;
};

export type AuthClientOptions = {
  id: string;
  loginTokenExpiration?: number;
} & (
  | {
      privateKey: KeyUtils.Key;
      token?: never;
      passphrase?: never;
    }
  | {
      privateKey?: never;
      token: string;
      passphrase?: never;
    }
  | {
      privateKey?: never;
      token?: never;
      passphrase: string;
    }
);
