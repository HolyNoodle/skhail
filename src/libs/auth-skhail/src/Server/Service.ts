import { Middleware, SkhailError } from "@skhail/core";
import { AuthInfo, KeyUtils } from "@skhail/auth";
import { AuthMiddleware } from "@skhail/auth/dist/server";
import { IInternalCredentials, ISkhailAuthenticationService } from "../Shared";

export interface SkhailClientRepository {
  get(id: string): Promise<IInternalCredentials | undefined>;
  create?(client: IInternalCredentials): Promise<void>;
  delete(id: string): Promise<void>;
}

export type SkhailAuthenticationServiceOptions = {
  privateKey: KeyUtils.Key;
  repository?: SkhailClientRepository;
  loginTokenExpiration?: number;
  internalClients?: IInternalCredentials[];
};

export class SkhailAuthenticationService extends ISkhailAuthenticationService {
  static middlewares: Middleware<any, any>[] = [
    new AuthMiddleware<ISkhailAuthenticationService>({
      register: true,
      login: true,
      unregister: true,
    }),
  ];

  constructor(private options: SkhailAuthenticationServiceOptions) {
    super();
  }

  async unregister(clientInfo: { id: string }) {
    const { id: credentialId } = this.context?.data?.authInfo ?? {};

    if (!credentialId) {
      throw new SkhailError({
        name: "denied",
        message: "You must be logged in to perform this action",
      });
    }

    if (clientInfo.id === credentialId) {
      throw new SkhailError({
        name: "denied",
        message: "Can't unregister your own credentials",
      });
    }

    await this.options.repository?.delete(clientInfo.id);

    await this.network.emit("deleted", [clientInfo.id]);
  }

  async register(clientInfo: { id: string; publicKey: string }): Promise<void> {
    if (!this.options.repository?.create) {
      throw new SkhailError({
        name: "denied",
        message: "Registration is disabled",
      });
    }

    const internalClient = this.options.internalClients?.find(
      (client) => client.id === clientInfo.id
    );
    const credentials =
      internalClient ?? (await this.options.repository.get(clientInfo.id));

    if (credentials) {
      throw new SkhailError({
        name: "denied",
        message: "Id already exists",
      });
    }

    const newCredentials: IInternalCredentials = {
      id: clientInfo.id,
      publicKey: clientInfo.publicKey,
    };

    await this.options.repository.create(newCredentials);

    await this.network.emit("created", [newCredentials.id]);
  }

  async login(loginInfo: { id: string; signedToken: string }): Promise<string> {
    const internalCred = this.options.internalClients?.find(
      (cred) => cred.id === loginInfo.id
    );
    const credentials =
      internalCred ?? (await this.options.repository?.get(loginInfo.id));

    if (!credentials) {
      throw new SkhailError({
        name: "denied",
        message: "Wrong credentials",
      });
    }

    const key = await KeyUtils.readKey(credentials.publicKey);
    await KeyUtils.verify<AuthInfo>(loginInfo.signedToken, key);

    return KeyUtils.sign(
      { id: credentials.id },
      this.options.privateKey,
      this.options.loginTokenExpiration ?? 5000
    );
  }
}
