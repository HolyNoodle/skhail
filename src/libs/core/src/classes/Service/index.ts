export { BatchService } from "../Helpers/BatchService";
export { EventConsumerService } from "../Helpers/EventConsumerService";
export { SkhailService } from "./Service";
