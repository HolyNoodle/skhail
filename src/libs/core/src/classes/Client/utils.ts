import {
  Constructor,
  ContextOptions,
  IEnveloppe,
  ILogger,
  IQueue,
  IRequestContext,
  ServiceFunctions,
} from "../../types";
import { SkhailError } from "../Error";
import { v4 } from "uuid";
import { SkhailService } from "../Service";

function wrapServiceCall<Context extends ContextOptions>(
  queue: IQueue<Context>,
  service: string,
  method: string,
  logger: ILogger,
  context?: Context,
  forwardedContext?: IRequestContext<Context>
): (...any: any[]) => Promise<any> {
  return async function serviceCall(...args) {
    const tid = forwardedContext?.tid || v4();

    logger.setTransactionId(tid);
    logger.trace("Service call", {
      service,
      method,
    });
    const enveloppe: IEnveloppe<Context> = {
      service,
      method,
      args,
      context: {
        tid,
        data: context || forwardedContext?.data,
        parent: forwardedContext,
      },
    };

    logger.debug("Enqueue enveloppe");
    const response = await queue.enqueue(enveloppe);
    logger.trace("Service response received", {
      service,
      method,
    });

    if (response.success === false) {
      throw new SkhailError(response.error);
    }

    return response.response;
  };
}

export function createServiceProxy<
  Service extends SkhailService<Context>,
  Context extends ContextOptions
>(
  type: Constructor<Service>,
  logger: ILogger,
  queue: IQueue<Context>,
  context?: Context,
  forwardedContext?: IRequestContext<Context>
): ServiceFunctions<Service, Context> {
  return new Proxy({} as ServiceFunctions<Service, Context>, {
    get(_, method: string) {
      return wrapServiceCall(
        queue,
        type.identifier,
        method,
        logger,
        context,
        forwardedContext
      );
    },
  });
}
