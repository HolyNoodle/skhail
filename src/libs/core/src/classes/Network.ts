import {
  Constructor,
  ContextOptions,
  EventFunction,
  EventOptions,
  IEmitter,
  IEventEmitter,
  IEventListener,
  ILogger,
  IQueue,
  QueueWrapper,
  ServiceFunctions,
} from "../types";
import { SkhailClient } from "./Client";
import { createServiceProxy } from "./Client/utils";
import { SkhailService } from "./Service";

export class SkhailNetwork<
  ContextType extends ContextOptions,
  EventType extends EventOptions
> implements
    IEventListener,
    IEventEmitter<EventType>,
    QueueWrapper<ContextType>
{
  constructor(
    private owner: SkhailService<any, EventType>,
    private logger: ILogger,
    private client: SkhailClient<ContextType>,
    public queue: IQueue<ContextType>,
    public pubsub?: IEmitter,
    private forwardedContext?: any
  ) {}

  private buildEvent(emitterService: Constructor<any>, event: string) {
    return `${emitterService.identifier}:${event}`;
  }

  on<
    Service extends SkhailService<any, EventType>,
    EventType extends EventOptions,
    Event extends keyof EventType
  >(
    service: Constructor<Service>,
    event: Event,
    listener: EventFunction<EventType[Event]>
  ): Promise<void> {
    if (!this.pubsub) {
      this.logger.warning("Event emitter not configured", {
        service: service.identifier,
        event,
      });

      return Promise.resolve();
    }

    this.owner.getLogger().trace("Service event start listening", {
      service: service.identifier,
      event,
    });
    return this.pubsub.on(
      (this.owner.constructor as Constructor<any>).identifier,
      this.buildEvent(service, event as string),
      listener as any
    );
  }

  off<
    Service extends SkhailService<any, EventType>,
    EventType extends EventOptions,
    Event extends keyof EventType
  >(
    service: Constructor<Service>,
    event: Event,
    listener: EventFunction<EventType[Event]>
  ): Promise<void> {
    if (!this.pubsub) {
      this.logger.warning("Event emitter not configured", {
        service: service.identifier,
        event,
      });

      return Promise.resolve();
    }

    this.owner.getLogger().trace("Service event stop listening", {
      service: service.identifier,
      event,
    });

    return this.pubsub.off(
      (this.owner.constructor as Constructor<any>).identifier,
      this.buildEvent(service, event as string),
      listener as any
    );
  }

  emit<Event extends keyof EventType>(
    event: Event,
    args: EventType[Event]
  ): Promise<void> {
    if (!this.pubsub) {
      this.logger.warning("Event emitter not configured", {
        service: (this.owner.constructor as Constructor<any>).identifier,
        event,
      });

      return Promise.resolve();
    }

    this.owner.getLogger().trace("Service event emition", {
      service: (this.owner.constructor as Constructor<any>).identifier,
      event,
    });
    return this.pubsub.emit(
      this.buildEvent(
        this.owner.constructor as Constructor<any>,
        event as string
      ),
      args
    );
  }

  get<Service extends SkhailService<ContextType, {}>>(
    type: Constructor<Service>,
    context?: ContextType | undefined
  ): ServiceFunctions<Service, ContextType> {
    return createServiceProxy(
      type,
      this.logger.clone(),
      this.queue,
      context,
      this.forwardedContext
    );
  }

  getClient() {
    return this.client;
  }
}
