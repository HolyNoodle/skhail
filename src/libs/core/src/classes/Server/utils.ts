import { SkhailServer } from ".";
import {
  Constructor,
  ContextOptions,
  IEnveloppe,
  IEnveloppeResponse,
  ILogger,
} from "../../types";
import { getError, SkhailError } from "../Error";
import { SkhailNetwork } from "../Network";
import { SkhailService } from "../Service";

function createServiceCopy<Service>(service: Service): Service {
  const executionCopy = Object.assign(
    Object.create(Object.getPrototypeOf(service)),
    service
  ) as Service;

  return executionCopy;
}

export function createLoggerCopy(
  logger: ILogger,
  service: Constructor,
  method?: string
): ILogger {
  const loggerCopy = logger.clone();

  loggerCopy.setScope(
    method ? [service.identifier, method].join(":") : service.identifier
  );

  return loggerCopy;
}

function getFunction(service: SkhailService<any>, method: string) {
  const methodRef = service[method as keyof SkhailService<any>];

  if (!methodRef) {
    throw new SkhailError({
      name: "not_found",
      message: "Method not found",
      details: {
        service: (service.constructor as Constructor).identifier,
        method,
      },
    });
  }

  return methodRef;
}

export function handleServiceCall<
  Service extends SkhailService<ContextType>,
  ContextType extends ContextOptions = ContextOptions
>(
  service: Service,
  server: SkhailServer<ContextType>,
  logger: ILogger
): (enveloppe: IEnveloppe<ContextType>) => Promise<IEnveloppeResponse> {
  return async (originalEnveloppe) => {
    const enveloppe = {
      ...originalEnveloppe,
      context: { ...originalEnveloppe.context },
    };

    const serviceCopy = createServiceCopy(service);
    const classObject = service.constructor as Constructor<SkhailService<any>>;

    const loggerCopy = createLoggerCopy(logger, classObject, enveloppe.method);
    loggerCopy.setTransactionId(enveloppe.context.tid);

    serviceCopy.setLogger(loggerCopy);
    const network = new SkhailNetwork(
      serviceCopy,
      loggerCopy,
      server.getClient(),
      server.getQueue(),
      server.getEmitter(),
      enveloppe.context
    );
    serviceCopy.setNetwork(network);

    const middlewares = classObject.getMiddlewares();
    if (middlewares) {
      for (let i = 0; i < middlewares.length; ++i) {
        const middleware = middlewares[i];

        try {
          const outScopeContext = await middleware.process(
            { ...enveloppe },
            service
          );
          enveloppe.context = {
            ...enveloppe.context,
            data: {
              ...(enveloppe.context.data ?? {}),
              ...(outScopeContext ?? {}),
            },
          };
        } catch (e) {
          const error = getError(
            e,
            {
              service: (service.constructor as Constructor).identifier,
              method: enveloppe.method,
              tid: enveloppe.context.tid,
            },
            {
              name: "unexpected",
              message: "An error occured while processing middlewares",
            }
          ).toObject();

          loggerCopy.error("Service call error", error);

          return {
            tid: enveloppe.context.tid,
            success: false,
            error,
          };
        }
      }
    }

    serviceCopy.setContext({ ...enveloppe.context });

    try {
      const method = getFunction(serviceCopy, enveloppe.method);

      loggerCopy.trace("Service execute");
      const result = await Reflect.apply(
        method as any,
        serviceCopy,
        enveloppe.args
      );
      loggerCopy.trace("Send service response");

      return {
        tid: enveloppe.context.tid,
        success: true,
        response: result,
      };
    } catch (e) {
      const error = getError(e, {
        service: enveloppe.service,
        method: enveloppe.method,
      }).toObject();

      loggerCopy.error("Service call error", error);

      return {
        tid: enveloppe.context.tid,
        success: false,
        error,
      };
    }
  };
}
