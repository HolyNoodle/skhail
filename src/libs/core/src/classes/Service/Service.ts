import {
  IRequestContext,
  ILogger,
  EventOptions,
  ContextOptions,
  Middleware,
  Constructor,
  ServiceFunctions,
} from "../../types";
import { SkhailNetwork } from "../Network";

export abstract class SkhailService<
  ContextType extends ContextOptions = ContextOptions,
  EventType extends EventOptions = {}
> {
  constructor() {
    if (!(this.constructor as any).identifier) {
      throw new Error(
        "SkhailService needs to implement static property 'identifier'."
      );
    }
  }

  protected static identifier: string;
  protected static alwaysManaged: boolean = false;
  protected static middlewares: Middleware<any>[] = [];

  ready?(): Promise<void>;
  prepare?(): Promise<void>;
  cleanup?(): Promise<void>;

  protected logger!: ILogger;
  protected context!: IRequestContext<ContextType>;
  protected network!: SkhailNetwork<ContextType, EventType>;

  static getMiddlewares() {
    return this.middlewares;
  }

  static isManaged() {
    return this.alwaysManaged;
  }

  getLogger() {
    return this.logger;
  }

  setLogger(logger: ILogger) {
    this.logger = logger;
  }

  setNetwork(network: SkhailNetwork<ContextType, EventType>) {
    this.network = network;
  }

  setContext(context: IRequestContext<ContextType>) {
    this.context = context;
  }

  get<Service extends SkhailService<ContextType, {}>>(
    type: Constructor<Service>,
    context?: ContextType | undefined
  ): ServiceFunctions<Service, ContextType> {
    return this.network.get(type, context);
  }
}
