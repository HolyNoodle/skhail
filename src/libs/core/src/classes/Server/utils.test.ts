/**
 * @group unit
 */
import { IEnveloppe } from "../../types";
import { SkhailNetwork } from "../Network";
import { SkhailService } from "../Service";
import { createLoggerCopy, handleServiceCall } from "./utils";

jest.mock("../Network");

class OtherTestService extends SkhailService {
  static identifier = "OtherTest";

  async other() {}
}

class TestService extends SkhailService {
  static identifier = "Test";

  spy: any;
  testMethod: any;
  internalTest() {
    (this as any)["spy"](this);
    return Promise.resolve();
  }

  nestedTest() {
    return this.get(OtherTestService).other();
  }
}

describe("handleServiceCall", () => {
  const queue = {
    props: "queue",
    enqueue: jest.fn().mockResolvedValue({
      success: true,
      response: undefined,
    }),
  } as any;
  const emitter = {
    props: "emitter",
  } as any;

  const server = {
    getQueue: jest.fn().mockReturnValue(queue),
    getEmitter: jest.fn().mockReturnValue(emitter),
    getClient: jest.fn().mockReturnValue({ prop: "client" }),
  } as any;

  const logger = {
    debug: jest.fn(),
    error: jest.fn(),
    info: jest.fn(),
    trace: jest.fn(),
    setTransactionId: jest.fn(),
    clone: jest.fn().mockReturnThis(),
    setScope: jest.fn(),
  } as any;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe("handleServiceCall", () => {
    it("Should call service method", async () => {
      const service = new TestService();
      const expectedResult = { object: "test" };
      const spy = jest.fn().mockReturnValue(expectedResult);
      service["logger"] = logger;

      (service as any)["testMethod"] = spy;

      const wrappedMethod = handleServiceCall(service, server, logger);

      const result = await wrappedMethod({
        service: TestService.identifier,
        method: "testMethod",
        args: ["truc", "muche"],
        context: { tid: "test transaction id" },
      });

      expect(result).toStrictEqual({
        tid: "test transaction id",
        success: true,
        response: expectedResult,
      });
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith("truc", "muche");
      expect(server.getQueue).toHaveBeenCalledTimes(1);
      expect(server.getEmitter).toHaveBeenCalledTimes(1);
      expect(SkhailNetwork).toHaveBeenCalledTimes(1);
      expect(SkhailNetwork).toHaveBeenCalledWith(
        expect.any(TestService),
        logger,
        { prop: "client" },
        queue,
        emitter,
        {
          tid: "test transaction id",
        }
      );
      expect(logger.trace).toHaveBeenCalledTimes(2);
      expect(logger.trace).toHaveBeenNthCalledWith(1, "Service execute");
      expect(logger.trace).toHaveBeenNthCalledWith(2, "Send service response");
    });

    it("Should forward context to server call", async () => {
      const service = new TestService();
      service.setLogger(logger);

      (SkhailNetwork.prototype.get as any).mockReturnValue({
        other: jest.fn().mockResolvedValue("other response"),
      });

      const wrappedMethod = handleServiceCall(service, server, logger);

      const result = await wrappedMethod({
        service: TestService.identifier,
        method: "nestedTest",
        args: [],
        context: { tid: "test transaction id" },
      });

      expect(result).toStrictEqual({
        tid: "test transaction id",
        success: true,
        response: "other response",
      });
      expect(SkhailNetwork.prototype.get).toHaveBeenCalledTimes(1);
      expect(SkhailNetwork.prototype.get).toHaveBeenCalledWith(
        OtherTestService,
        undefined
      );
    });

    it("Should copy context on service", async () => {
      const network = {
        queue: {
          enqueue: jest.fn(),
        },
      } as any;
      const server = {
        get: jest.fn().mockReturnValue({ other: jest.fn() }),
        getQueue: jest.fn(),
        getEmitter: jest.fn(),
        getClient: jest.fn().mockReturnValue({}),
      } as any;
      const logger = {
        error: jest.fn(),
        info: jest.fn(),
        trace: jest.fn(),
        setTransactionId: jest.fn(),
        clone: jest.fn().mockReturnThis(),
        setScope: jest.fn(),
      } as any;
      const service = new TestService();
      const expectedResult = { object: "test" };
      const spy = jest.fn().mockReturnValue(expectedResult);
      service["network"] = network;
      service["logger"] = logger;

      (service as any)["testMethod"] = spy;

      const wrappedMethod = handleServiceCall(service, server, logger);

      const result = await wrappedMethod({
        service: TestService.identifier,
        method: "testMethod",
        args: ["truc", "muche"],
        context: { tid: "test transaction id" },
      });

      expect(result).toStrictEqual({
        tid: "test transaction id",
        success: true,
        response: expectedResult,
      });
      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith("truc", "muche");
    });

    it("Should not call service method when method does not exist", async () => {
      const network = {
        queue: {
          enqueue: jest.fn(),
        },
      } as any;
      const server = {
        get: jest.fn().mockReturnValue({ other: jest.fn() }),
        getQueue: jest.fn(),
        getEmitter: jest.fn(),
        getClient: jest.fn().mockReturnValue({}),
      } as any;
      const logger = {
        error: jest.fn(),
        info: jest.fn(),
        trace: jest.fn(),
        setTransactionId: jest.fn(),
        clone: jest.fn().mockReturnThis(),
        setScope: jest.fn(),
      } as any;
      const service = new TestService();
      const spy = jest.fn();
      service["network"] = network;
      service["logger"] = logger;

      (service as any)["testMethod"] = spy;

      const wrappedMethod = handleServiceCall(service, server, logger);

      const result = await wrappedMethod({
        service: TestService.identifier,
        method: "testMethodNotExist",
        args: ["truc", "muche"],
        context: { tid: "test transaction id" },
      });

      expect(result).toStrictEqual({
        tid: "test transaction id",
        success: false,
        error: {
          details: {
            method: "testMethodNotExist",
            service: "Test",
          },
          message: "Method not found",
          name: "not_found",
        },
      });
    });

    it("Should copy the service", async () => {
      const network = {
        server: "test",
        getQueue: jest.fn(),
        getEmitter: jest.fn(),
      } as any;
      const server = {
        get: jest.fn().mockReturnValue({ other: jest.fn() }),
        getQueue: jest.fn(),
        getEmitter: jest.fn(),
        getClient: jest.fn().mockReturnValue({}),
      } as any;
      const logger = {
        error: jest.fn(),
        info: jest.fn(),
        trace: jest.fn(),
        setTransactionId: jest.fn(),
        clone: jest.fn().mockReturnThis(),
        setScope: jest.fn(),
      } as any;
      const service = new TestService();
      const spy = jest.fn();
      service["network"] = network;
      service["logger"] = logger;

      (service as any)["spy"] = spy;

      const wrappedMethod = handleServiceCall(service, server, logger);

      await wrappedMethod({
        service: TestService.identifier,
        method: "internalTest",
        args: ["truc", "muche"],
        context: { tid: "test transaction id" },
      });

      expect(spy).toHaveBeenCalledTimes(1);

      const serviceCopy = spy.mock.calls[0][0];

      expect(serviceCopy).not.toBe(service);
      expect(serviceCopy).toBeInstanceOf(TestService);
      expect(serviceCopy.logger).toBe((service as any).logger);
      expect(serviceCopy.context).toStrictEqual({ tid: "test transaction id" });
      expect(serviceCopy.server).toStrictEqual((service as any).server);
    });

    it("Should run middlewares in chain", async () => {
      const network = {
        server: "test",
        getQueue: jest.fn(),
        getEmitter: jest.fn(),
      } as any;
      const server = {
        server: "test",
        get: jest.fn(),
        getQueue: jest.fn(),
        getEmitter: jest.fn(),
        getClient: jest.fn().mockReturnValue({}),
      } as any;
      const logger = {
        error: jest.fn(),
        info: jest.fn(),
        trace: jest.fn(),
        setTransactionId: jest.fn(),
        clone: jest.fn().mockReturnThis(),
        setScope: jest.fn(),
      } as any;
      const enveloppe: IEnveloppe<any> = {
        service: TestService.identifier,
        method: "internalTest",
        args: ["truc", "muche"],
        context: {} as any,
      };
      const middlewares = [
        {
          id: "mid1",
          process: jest.fn().mockResolvedValue({ id: "test", env: "1" }),
        },
        {
          id: "mid2",
          process: jest.fn().mockResolvedValue({ env: "2" }),
        },
        {
          id: "mid3",
          process: jest.fn().mockResolvedValue(undefined),
        },
      ];

      const service = new TestService();
      TestService["middlewares"] = middlewares;

      service["network"] = network;
      service["logger"] = logger;
      (service as any)["spy"] = jest.fn();

      const wrappedMethod = handleServiceCall(service, server, logger);

      await wrappedMethod(enveloppe);

      expect(middlewares[0].process).toHaveBeenCalledTimes(1);
      expect(middlewares[0].process).toHaveBeenCalledWith(
        {
          ...enveloppe,
          context: {},
        },
        service
      );
      expect(middlewares[1].process).toHaveBeenCalledTimes(1);
      expect(middlewares[1].process).toHaveBeenCalledWith(
        {
          ...enveloppe,
          context: {
            data: {
              env: "1",
              id: "test",
            },
          },
        },
        service
      );
      expect(middlewares[2].process).toHaveBeenCalledTimes(1);
      expect(middlewares[2].process).toHaveBeenCalledWith(
        {
          ...enveloppe,
          context: {
            data: {
              env: "2",
              id: "test",
            },
          },
        },
        service
      );
    });

    it("Should stop calling middle when one fails", async () => {
      const network = {
        server: "test",
        getQueue: jest.fn(),
        getEmitter: jest.fn(),
      } as any;
      const server = {
        server: "test",
        get: jest.fn(),
        getQueue: jest.fn(),
        getEmitter: jest.fn(),
        getClient: jest.fn().mockReturnValue({}),
      } as any;
      const logger = {
        error: jest.fn(),
        info: jest.fn(),
        trace: jest.fn(),
        setTransactionId: jest.fn(),
        clone: jest.fn().mockReturnThis(),
        setScope: jest.fn(),
      } as any;
      const enveloppe: IEnveloppe<any> = {
        service: TestService.identifier,
        method: "internalTest",
        args: ["truc", "muche"],
        context: { context: "test" } as any,
      };
      const middlewares = [
        {
          id: "mid1",
          process: jest.fn().mockRejectedValue("Error test"),
        },
        {
          id: "mid2",
          process: jest.fn().mockResolvedValue({ env: "2" }),
        },
      ];

      const service = new TestService();
      TestService["middlewares"] = middlewares;
      service["network"] = network;
      service["logger"] = logger;
      (service as any)["spy"] = jest.fn();

      const wrappedMethod = handleServiceCall(service, server, logger);
      const result = await wrappedMethod(enveloppe);

      expect(middlewares[0].process).toHaveBeenCalledTimes(1);
      expect(middlewares[1].process).toHaveBeenCalledTimes(0);

      expect(result.success).toBeFalsy();
      expect((result as any).error).toMatchSnapshot();
    });

    it("Should not fail if middlewares are undefined", async () => {
      const network = {
        server: "test",
        getQueue: jest.fn(),
        getEmitter: jest.fn(),
      } as any;
      const server = {
        server: "test",
        get: jest.fn(),
        getQueue: jest.fn(),
        getEmitter: jest.fn(),
        getClient: jest.fn().mockReturnValue({}),
      } as any;
      const logger = {
        error: jest.fn(),
        info: jest.fn(),
        trace: jest.fn(),
        setTransactionId: jest.fn(),
        clone: jest.fn().mockReturnThis(),
        setScope: jest.fn(),
      } as any;
      const enveloppe: IEnveloppe<any> = {
        service: TestService.identifier,
        method: "internalTest",
        args: ["truc", "muche"],
        context: { context: "test" } as any,
      };

      const service = new TestService();
      TestService["middlewares"] = undefined as any;
      service["network"] = network;
      service["logger"] = logger;
      (service as any)["spy"] = jest.fn();

      const wrappedMethod = handleServiceCall(service, server, logger);
      await wrappedMethod(enveloppe);

      expect((service as any)["spy"]).toHaveBeenCalledTimes(1);
      expect(logger.setScope).toHaveBeenCalledTimes(1);
      expect(logger.setScope).toHaveBeenCalledWith("Test:internalTest");
    });
  });

  describe("createLoggerCopy", () => {
    it("Should clone and set service on logger", () => {
      const logger = {
        error: jest.fn(),
        info: jest.fn(),
        trace: jest.fn(),
        setTransactionId: jest.fn(),
        setService: jest.fn(),
        clone: jest.fn().mockReturnThis(),
        setScope: jest.fn(),
      } as any;
      const loggerCopy = createLoggerCopy(logger, TestService);

      expect(loggerCopy).toBe(logger);
      expect(logger.setScope).toHaveBeenCalledTimes(1);
      expect(logger.setScope).toHaveBeenCalledWith(TestService.identifier);
    });
  });
});
