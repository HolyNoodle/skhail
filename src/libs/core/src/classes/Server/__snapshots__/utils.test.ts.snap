// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleServiceCall handleServiceCall Should stop calling middle when one fails 1`] = `
{
  "details": {
    "error": {
      "message": "Error test",
    },
    "method": "internalTest",
    "service": "Test",
    "tid": undefined,
  },
  "message": "An error occured while processing middlewares",
  "name": "unexpected",
}
`;
